[ImagePopout] Update types to 0.7.9

Takitaf opened the pull request:
Closes #59 

I changed `O` generic in `FormApplication` to allow primitives as constructor parameter (as suggested by BoltsJ in #59).

Please, let me know if the way I handled the first generic (`D`) is a proper way of doing that. As far as I can see, the `D` generic of `FormApplication` is mostly used for returning type of `getData` method, so I've created `ImagePopout.Data` type with additional fields (pulled out from `ImagePopout.getData` source code). 

Maybe I should make `D` a little bit more strict for `ImagePopoutData` and change it to `D extends ImagePopout.Data = ImagePopout.Data`?

Also note that, according to source code, `_updateObject` is not overridden in `ImagePopout`. Unfortunately, `FormApplication._updateObject` is abstract, so it forces us to `implement it` in some way. In foundry code, it's just method with single exception thrown, so I made `_updateObject` with `never` return type. If you have a pattern for that, let me know.

(I did `npm run lint` before push and there were no errors in changed files)

BoltsJ commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+> {
+  constructor(src: O, options?: ImagePopout.Options);
+
+  protected _related: null;
```
This should be `Entity | object | null`. Possibly even more specific.

BoltsJ commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  protected static _handleShareImage(params: { image: string; title: string; uuid: string }): ImagePopout;
+
+  /** @override */
+  protected _updateObject(event: Event, formData?: object): never;
```
You should add `@remarks Not implemented for ImagePopout` to the docstring here.

Takitaf commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
  */
-declare class ImagePopout extends FormApplication {
-  _updateObject(event?: Event, formData?: object): Promise<any>; // TODO: Add correct return type
+declare class ImagePopout<D extends object = ImagePopout.Data, O extends string = string> extends FormApplication<
```
Probably not. I can get rid of them.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
````diff
@@ -1,10 +1,96 @@
+ * Furthermore, this application allows for sharing the display of an image with other connected players.
+ *
+ * @example
+ * ```javascript
````
The code snippet should work just as well in Typescript, so we can just as well mark it as typescript.
```suggestion
 * ```typescript
```

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  /**
+   * @override
+   * @remarks Not implemented for ImagePopout
+   * */
```
```suggestion
   */
```

kmoschcau reviewed (changes requested):
I have a few formatting changes, I am going to commit them right here on GitHub.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+}
+
+declare namespace ImagePopout {
+  interface Options extends FormApplication.Options {
+    shareable: boolean;
+    uuid: string | null;
+  }
```
This should not only add the new options, but should also relist the already existing options with their explicitly defined `@defaultValue`s documented.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  get title(): string;
+
+  /** @override */
+  getData(options?: Application.RenderOptions): ImagePopout.Data | Promise<ImagePopout.Data>;
```
This always returns a Promise as far as I can tell, so the return type can be narrowed.
```suggestion
  getData(options?: Application.RenderOptions): Promise<ImagePopout.Data>;
```

ghost91- replied:
I think the point of having the return type this wide is to allow people who derive from this class implement it in the way that they want. We did that in a couple of other places. But might not make sense here.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  /**
+   * Determine the correct position and dimensions for the displayed image
+   */
+  protected static getPosition(img: string): ImagePopout.Position;
```
This is not entirely correct. The first line in the function causes this to return an object with only `width` and `height`. I suggest using a union type here and also having a look at `Application.Position`.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  getRelatedObject(): Promise<Entity | object | null>;
+
+  /** @override */
+  protected _render(force?: boolean, options?: Application.RenderOptions): Promise<void>;
```
This method sets `this.position` with the return type of `getPosition`, which might not be a full `Application.Position`. That should be explicitly typed and if need be, changed in `Application`.

Takitaf replied:
Well, I found out that I need to to the same on my previous class (journalSheet), as it uses that `getPosition` inside and also sets its `this.position` to that.

I'm not sure how many times the `Application.Position` will be used throughout the application - maybe `scale` property can be marked as optional. How about sticking with:
```typescript
type ImagePopuout.Position = {width: number; height: number; top: number; left: number;} | {width: number; height: number}
```
with `todo` to check at the end of 0.7.9 migration if we can mark `scale` as optional and put `Application.Position` as first parameter in union above. What do you think, @kmoschcau?

kmoschcau replied:
At that point I would just make `top`, `left` and `scale` of `Application.Position` optional and use that everywhere. If it can then be said for a return value that the properties are definitely said in one case, you can use `Required<Application.Position>` for example. @ghost91- @BoltsJ what do you guys think?

ghost91- replied:
I think `scale` can always be optional, so using `Required` on the return type of `setPosition` would not work:
```js
game.actors.getName('Gorgoth').sheet.setPosition({height: 'auto'})
{width: 745, height: 1106, left: 185.5, top: 23, scale: undefined}
```

Takitaf replied:
In that case, maybe I just make the `Application.Position.scale` optional and stick with
```typescript
type ImagePopout.Position = Application.Position | {width: number, height: number}
```
The question is how many other classes have that `incomplete position` inside. I can see that it is the case in both `ImagePopout` and `JournalSheet`. Maybe that type above is not a bad idea, as it can be just an exception to the rule. @kmoschcau @ghost91- @BoltsJ ?

ghost91- replied:
@Takitaf I'd need to take a closer look at this first to completely understand what is going on. Does it make sense for me to dive deeply into this right now or will you be able to handle this without me? :D What is the state of this PR in general? Just want to make sure my (already quite limited) time is well invested :)

Takitaf replied:
@ghost91- Well, I think everything is done on my local environment, except this one. 

As far as I can see (I don't know whole codebase well, at least for now, unfortunately), this class has narrower `position` field type (only `width` and `height` are set). Either we can change the `Application.Position` type by making `scale`, `top` and `left` properties optional, or we can make union type on `position` field `Application.Position | {width: number, height: number}`. Both changes can potentially make some problems in a future probably.

Anyway, I wanted to commit cr fixes, but someone updated `husky` lib and it seems to not working properly (at least on my local env with Windows), so I need to find a solution to that in order to be able to commit changes.

ghost91- replied:
OK, I will take a look then. We should be able to find a solution. It would be great if we coud move forward with this.

Regarding your issue with husky, feel free to ping me in the League Discord (I'm called "ghost" there), maybe we can figure it out together.

kmoschcau replied:
You could do it like this `Pick<Application.Position, 'width' | 'height'>`, or even this `Partial<Application.Position> & Pick<Application.Position, 'width' | 'height'>`.

kmoschcau commented on foundry/applications/formApplications/imagePopout.d.ts:
```diff
@@ -1,10 +1,96 @@
+  /**
+   * Handle a received request to display an image.
+   */
+  protected static _handleShareImage(params: { image: string; title: string; uuid: string }): ImagePopout;
```
This should be changed to stick with the conventions up to now. It will likely require some formatting in addition.
```suggestion
  protected static _handleShareImage({ image, title, uuid }: { image: string; title: string; uuid: string }): ImagePopout;
```

Takitaf commented:
Thanks @kmoschcau, I'll fix this today's evening.

kmoschcau commented:
Don't forget to pull first. :D

Takitaf commented:
@kmoschcau I've fixed all CR issues you've mentioned.

@ghost91- @kmoschcau - According to `position` typing issue - I just changed it to `position: Application.Position | Pick<Application.Position, 'width' | 'height'>;` typing in `Application`. I guess it can be still far from ideal, but maybe we can finish this PR and put the whole `Application.Position` problem in its own issue? (as it potentially is connected with few others classes as well).

If you accept this, I would go back to `JournalSheet` (my first PR), because you mentioned few issues in this PR that are probably wrong in `JournalSheet` as well but eh, still learning convention and codebase. ;)

ghost91- commented:
Sounds reasonable to me. Could you create that issue?
